walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,150 @@
+    int numEvents = events.size();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      TimeRange when = event.getWhen();
+      int start = (when.start()) / 30;
+      int eventSlots = when.duration() / 30;
+
+      Iterator<String> eventAi = eventAttendees.iterator(); 
+      while(eventAi.hasNext()) {
+        numEvents--;
+        String eventAttendee = eventAi.next();
+
+        Iterator<String> requestAi = requestAttendees.iterator(); 
+        while (requestAi.hasNext()) {
+          String requestAttendee = requestAi.next();
+          if (eventAttendee.equals(requestAttendee)) {
+            for (int i = 0; i < eventSlots; i++) 
+              timeSlots[start + i] = false;
+            break;
+          }
+        }
+        if(numEvents == 0)
+            break;
+      }
+    }
+    return timeSlots;
+  }
+
+  // converts boolean availability to list of available slots
+  public static ArrayList<Integer> convertToSlots (boolean timeSlots[], int numRequestSlots) {
+    ArrayList<Integer> slots = new ArrayList<Integer>();
+    boolean add;
+    for (int i = 0; i < 48; i++) {
+      add = false;
+      if (timeSlots[i]) {
+        add = true;
+        // checks if the duration is longer than a single time slot
+        for (int j = 1; j < numRequestSlots; j++)
+          if ((i+j) < timeSlots.length && !timeSlots[i+j])
+            add = false;
+      }
+      if(add)
+        for (int j = 0; j < numRequestSlots; j++)
+          slots.add(i+j);
+    }
+    return slots;
+  }
+
+  // converts list of available slots to a list of TimeRanges 
+  public static ArrayList<TimeRange> convertToTimes (ArrayList<Integer> slots, int numRequestSlots, int numEvents) {
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    int start;
+    boolean done = false;
+
+    // if slots size is zero
+    if(slots.size() == 0) 
+      return possibleTimes;
+
+    start = slots.get(0)*30;
+
+    // if there's just enough room
+    if (numRequestSlots == slots.size()){
+      possibleTimes.add(TimeRange.fromStartEnd(start, start + numRequestSlots*30, false));
+      return possibleTimes;
+    }
+
+    // if the whole day is available
+    if (slots.size() == 48){
+      possibleTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+      return possibleTimes;
+    }
+
+    // everything else
+    for (int i = 0; i < slots.size() - numRequestSlots; i++) {
+      // if the next value at the next index is not +1 from the last
+      int current = slots.get(i);
+      int next = slots.get(i+1);
+      if (current != next - 1) {
+        possibleTimes.add(TimeRange.fromStartEnd(start, current*30+30, false));
+        start = next*30;
+        numEvents--;
+      }
+      if (i == slots.size() - numRequestSlots - 1)
+        done = true;
+      if (numEvents == 0){
+        done = true;
+        break;
+      }
+    }
+    if(done) {
```

Comment:
Will done ever be false? It seems like it will always be set to true in the above loop